src/solvers/dgsem_p4est/dg_3d_parabolic.jl

```diff
@@ -0,0 +1,717 @@
+                                  Ja32 * gradients_reference_3
+                gradient_z_node = Ja13 * gradients_reference_1 +
+                                  Ja23 * gradients_reference_2 +
+                                  Ja33 * gradients_reference_3
+
+                set_node_vars!(gradients_x, gradient_x_node, equations_parabolic, dg, i, j,
+                               k,
+                               element)
+                set_node_vars!(gradients_y, gradient_y_node, equations_parabolic, dg, i, j,
+                               k,
+                               element)
+                set_node_vars!(gradients_z, gradient_z_node, equations_parabolic, dg, i, j,
+                               k,
+                               element)
+            end
+        end
+    end
+    # @show norm(gradients_y)
+    # Prolong solution to interfaces
+    @trixi_timeit timer() "prolong2interfaces" begin
+        prolong2interfaces!(cache_parabolic, u_transformed, mesh,
+                            equations_parabolic, dg.surface_integral, dg)
+    end
+
+    # Calculate interface fluxes for the gradient. This reuses P4est `calc_interface_flux!` along with a
+    # specialization for AbstractEquationsParabolic.
+    @trixi_timeit timer() "interface flux" begin
+        calc_interface_flux!(cache_parabolic.elements.surface_flux_values,
+                             mesh, False(), # False() = no nonconservative terms
+                             equations_parabolic, dg.surface_integral, dg, cache_parabolic)
+    end
+
+    # Prolong solution to boundaries
+    @trixi_timeit timer() "prolong2boundaries" begin
+        prolong2boundaries!(cache_parabolic, u_transformed, mesh,
+                            equations_parabolic, dg.surface_integral, dg)
+    end
+
+    # Calculate boundary fluxes
+    @trixi_timeit timer() "boundary flux" begin
+        calc_boundary_flux_gradients!(cache_parabolic, t, boundary_conditions_parabolic,
+                                      mesh, equations_parabolic, dg.surface_integral, dg)
+    end
+
+    # TODO: parabolic; mortars
+    @assert nmortars(dg, cache) == 0
+
+    # Calculate surface integrals
+    @trixi_timeit timer() "surface integral" begin
+        (; boundary_interpolation) = dg.basis
+        (; surface_flux_values) = cache_parabolic.elements
+        (; contravariant_vectors) = cache.elements
+
+        # Access the factors only once before beginning the loop to increase performance.
+        # We also use explicit assignments instead of `+=` to let `@muladd` turn these
+        # into FMAs (see comment at the top of the file).
+        factor_1 = boundary_interpolation[1, 1]
+        factor_2 = boundary_interpolation[nnodes(dg), 2]
+        @threaded for element in eachelement(dg, cache)
+            for m in eachnode(dg), l in eachnode(dg)
+                for v in eachvariable(equations_parabolic)
+                    for dim in 1:3
```

Comment:
Did you check whether this is the best order of the loops?

Reply:
Do you mean best in terms of memory access?

Reply:
Yes.

Reply:
What about this?

Reply:
Bump

Reply:
Hi @ranocha 
I timed the parabolic surface integral and realized that the following order of indexing is the fastest:

So the loop first access dim index, then the order is v, m , l and finally element.
                             
dim, v, m , l, element : 4.91k    8.02s   11.6%  1.63ms     0.00B    0.0%    0.00B
dim, v, l, m, element  : 4.91k    8.14s   11.7%  1.66ms     0.00B    0.0%    0.00B

Reply:
Are you sure this is not benchmarking noise and doesn't really make a difference? I was more wondering whether it could be beneficial to move the `for dim in 1:3` loop so that `gradients[dim]` is handled one `dim` after another?

Reply:
@ranocha,
I performed some testing and moved for dim in 1:3 to the outer loop and also other combinations. I ran each case twice and I use the "elixir_navierstokes_convergence.jl" for the benchmarking. The timing is not the benchmark noise and it does not change after I ran it many times. Here are the results. I list the indices based on the order of memory access. The best order is "dim, v, m, l, element"

dim, v, m, l , element:  1.63ms,1.63ms
v, m, l, dim
, element:   1.81ms,1.80ms
v, m, l, element, dim:   1.81ms,1.81ms
v, l, m, element, dim:   1.83ms, 1.84ms
dim, v, l, m, element:   1.65ms,1.65ms

Reply:
Thanks for checking! Sounds like moving `dim` to the outside loop is the most efficient.

Reply:
@threaded for element in eachelement(dg, cache)
            for l in eachnode(dg), m in eachnode(dg)
                for v in eachvariable(equations_parabolic)
                    for dim in 1:3

Reply:
dim should be the most inner loop. the ordering is based on memory access.

Reply:
Ah, I misunderstood. Thanks for clarifying

Reply:
No worries.

Reply:
Okay, thanks